# Shopify Summer 2022 Data Science Intern Challege

**Question 1**: Given some sample data, write a program to answer the following: 
On Shopify, we have exactly 100 sneaker shops, and each of these shops sells only one model of shoe. We want to do some analysis of the average order value (AOV). When we look at orders data over a 30 day window, we naively calculate an AOV of $3145.13. Given that we know these shops are selling sneakers, a relatively affordable item, something seems wrong with our analysis. 

1. Think about what could be going wrong with our calculation. Think about a better way to evaluate this data. 
2. What metric would you report for this dataset?
3. What is its value?


In [1]:
import numpy as np
import pandas as pd

In [2]:
# Reading Data
df=pd.read_csv("2019 Winter Data Science Intern Challenge Data Set - Sheet1.csv")
df.head()

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   order_id        5000 non-null   int64 
 1   shop_id         5000 non-null   int64 
 2   user_id         5000 non-null   int64 
 3   order_amount    5000 non-null   int64 
 4   total_items     5000 non-null   int64 
 5   payment_method  5000 non-null   object
 6   created_at      5000 non-null   object
dtypes: int64(5), object(2)
memory usage: 273.6+ KB


###### 1.Think about what could be going wrong with our calculation. Think about a better way to evaluate this data. <br>
Ans. The average order value is calculated as `sum(order_amount)/total number of orders` by forgetting the fact that an order can have multiple items. The better way to evaluate this data is to consider the `total_items` while calculating average.

In [4]:
# Wrong calculation of AOV
print("Average Order Value : ${:.2f}".format(df['order_amount'].mean()))

#which is same as
avg = sum(df['order_amount'])/df['order_amount'].count()
print(f"Detailed Calculation Average : ${avg:.2f}")

Average Order Value : $3145.13
Detailed Calculation Average : $3145.13


###### 2.What metric would you report for this dataset?
Ans. The AOV should be calculated as `sum(order_amount)/sum(total_items)` for the given dataset.

In [5]:
correct_aov = sum(df['order_amount'])/sum(df['total_items'])
print(f"The Correct Average Order Value = ${correct_aov:.2f}")

The Correct Average Order Value = $357.92


###### 3. What is its value?
Ans. The correct AOV is `$357.92` approx.

<hr>

**Question 2**: For this question you’ll need to use SQL. <a href="https://www.w3schools.com/SQL/TRYSQL.ASP?FILENAME=TRYSQL_SELECT_ALL">Follow this link</a> to access the data set required for the challenge. Please use queries to answer the following questions. Paste your queries along with your final numerical answers below.

**1. How many orders were shipped by Speedy Express in total?**<br>
```
SELECT COUNT(orderID) AS "Number of orders shipped by Speedy Express"
FROM Orders JOIN Shippers USING(ShipperID)
WHERE ShipperName="Speedy Express";
```

**Result:**<br>
Number of orders shipped by Speedy Express<br>
54

**2. What is the last name of the employee with the most orders?**
```
SELECT lastname AS "Last Name"
FROM (SELECT MAX(totalOrders),lastname 
	  FROM (SELECT COUNT(orderId) AS totalOrders, employeeID, lastname 
		    FROM ORDERS JOIN employees USING(employeeID) GROUP BY employeeID))
```
**Result:**<br>
Last Name<br>
Peacock

**3. What product was ordered the most by customers in Germany?**
```
SELECT productName AS "Product Name" 
FROM (SELECT MAX(Total), productName 
		FROM (SELECT productID, SUM(Quantity) AS Total, productName 
        	  FROM (SELECT * FROM Customers JOIN ( SELECT orderID,productID,Quantity,CustomerID 
                                                   FROM Orders JOIN OrderDetails USING (orderID)) 
                                            USING (customerID) WHERE LOWER(Country) LIKE "germany")
              JOIN products USING (productID) GROUP BY productID))
```
**Result:**<br>
Product Name<br>
Boston Crab Meat